Add cloudwatchLog event

horike37 opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:
A Lambda function is called by CloudWatchLogs Events
Closes #3399

<!--
Briefly describe the feature if no issue exists for this PR
-->

## How did you implement it:
I created cloudwatchlog directory in the following directory and implemented it.
https://github.com/serverless/serverless/tree/master/lib/plugins/aws/deploy/compile/events
<!--
If this is a nontrivial change please briefly describe your implementation so its easy for us to understand and review your code.
-->

## How can we verify it:
Please write in serverless.yml as follow, then run `sls deploy`

```yaml
events:
    - cloudwatchLog:
        logGroup: '/aws/lambda/hello'
        filter: '{$.userIdentity.type = Root}'
```

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* AWS CLI commands - To list AWS resources and show that the correct config is in place
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

horike37 commented:
I have something I would like to consult with @serverless/team .
I have implemented to directly specify `logGroup` name. 
However, if Lambda's log group managed by Serverless set as an event source, I think it would be better to have a way to refer to it.

For example as follow
```
functions:
  hello:
    handler: handler.hello
  log:
    handler: handler.log
    events:
        - cloudwatchLog:
            logGroup: ${loggroup:hello}
```

What do you think?

buggy commented on lib/plugins/aws/deploy/compile/events/cloudWatchLog/index.js:
```diff
@@ -0,0 +1,124 @@
+            if (typeof event.cloudwatchLog === 'object') {
+              if (!event.cloudwatchLog.logGroup) {
+                const errorMessage = [
+                  'Missing "logGroup" or "filter" property for logs event ',
```
I think it would be better if this said: `Missing "logGroup" property for cloudwatchLog event`. Having `or "filter"` implies that I can specify a `filter` option only which is not true.

pmuens replied:
Nice catch. Yes, I agree. The error message might be a little bit misleading

buggy commented:
Hey @horike37. I'm really glad to see this feature. I'm just wondering about the configuration. Comparing it to the implementations for SNS, S3, Schedules, etc:


1. They support a short form configuration. Is something like this also supported?

```yaml
    events:
        - cloudwatchLog: '/aws/lambda/hello'
```

2. Could I use an arn instead of a log group name?

```yaml
    events:
        - cloudwatchLog:
            logGroup: arn:xxx
```

or

```yaml
    events:
        - cloudwatchLog: arn:xxx
```

pmuens commented:
Hey @horike37 !

Thank you very much for implementing this! You've done it again! 👍 🥇 🎉 

I'll review it today!

> However, if Lambda's log group managed by Serverless set as an event source, I think it would be better to have a way to refer to it.

Yes, I agree. It would be nice to specify it. I think that using the `arn` as described by @buggy might be the best solution here since other event sources use this as well 👍 

Let us know if we can assist you!

pmuens commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -220,6 +220,12 @@ module.exports = {
+  // CloudWatch Log
+  getCloudWatchLogLogicalId(functionName, logsIndex) {
+    return `${this
+      .getNormalizedFunctionName(functionName)}CloudWatchLog${logsIndex}`;
```
This should be renamed from `{functionName}CloudWatchLogs{logsIndex}` to `{functionName}LogsSubscriptionFilterCloudWatchLog{logsIndex}`

since we're using the `AWS::Logs::SubscriptionFilter` resource.

This way it's more predictable for the user.

The `cloudwatchEvent` event e.g. uses a `AWS::Events::Rule` resource and therefore the name of the generated resource is `{normalizedFunctionName}EventsRuleCloudWatchEvent{sequentialId}`

See: https://serverless.com/framework/docs/providers/aws/guide/resources/

pmuens commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -248,4 +254,8 @@ module.exports = {
     return `${this.getNormalizedFunctionName(functionName)}LambdaPermissionAlexaSkill`;
   },
+  getLambdaCloudWatchLogPermissionLogicalId(functionName, logsIndex) {
+    return `${this.getNormalizedFunctionName(functionName)}LambdaPermissionCloudWatchLog${
```
This should be renamed to `{functionName}LambdaPermissionLogsSubscriptionFilterCloudWatchLog{logsIndex}` so that it's consistent with the `cloudwatchEvent` event permission logical id (Line 246 in this file).

See: https://serverless.com/framework/docs/providers/aws/guide/resources/

andymac4182 commented:
It would be great if you could specify `*` so you can have 1 function that receives all loggroups from the service. We are doing this currently to push our lambda logs to our logging server.

pmuens commented on docs/providers/aws/guide/resources.md:
```diff
@@ -70,8 +70,9 @@ We're also using the term `normalizedName` or similar terms in this guide. This
 |Lambda::Version        | {normalizedFunctionName}LambdaVersion{sha256}                           | HelloLambdaVersionr3pgoTvv1xT4E4NiCL6JG02fl6vIyi7OS1aW0FwAI |
 |Logs::LogGroup         | {normalizedFunctionName}LogGroup                        | HelloLogGroup                 |
-|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{index} </li><li>**S3**: {normalizedFunctionName}LambdaPermission{normalizedBucketName}S3</li><li>**APIG**: {norm ...
+|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**CloudWatch Log**: {normalizedFunctionName}LambdaPermissionCloudWatchLog{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{index} </li><li>**S3**: ...
```
This part of the table needs an update after the logicalIds have been renamed.

pmuens commented on docs/providers/aws/guide/resources.md:
```diff
@@ -70,8 +70,9 @@ We're also using the term `normalizedName` or similar terms in this guide. This
-|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{index} </li><li>**S3**: {normalizedFunctionName}LambdaPermission{normalizedBucketName}S3</li><li>**APIG**: {norm ...
+|Lambda::Permission     | <ul><li>**Schedule**: {normalizedFunctionName}LambdaPermissionEventsRuleSchedule{index}</li><li>**CloudWatch Event**: {normalizedFunctionName}LambdaPermissionEventsRuleCloudWatchEvent{index}</li><li>**CloudWatch Log**: {normalizedFunctionName}LambdaPermissionCloudWatchLog{index}</li><li>**IoT**: {normalizedFunctionName}LambdaPermissionIotTopicRule{index} </li><li>**S3**: ...
 |Events::Rule           | <ul><li>**Schedule**: {normalizedFuntionName}EventsRuleSchedule{SequentialID}</li><li>**CloudWatch Event**: {normalizedFuntionName}EventsRuleCloudWatchEvent{SequentialID}</li> </ul> | <ul><li>**Schedule**: HelloEventsRuleSchedule1</li><li>**CloudWatch Event**: HelloEventsRuleCloudWatchEvent1</li></ul>      |
+|AWS::Logs::SubscriptionFilter    | {normalizedFuntionName}CloudWatchLog{SequentialID}       | HelloCloudWatchLog1            |
```
Same here.

HyperBrain commented:
It is also important, that a log stream can have one subscription at maximum. This should be checked by the code, as otherwise the CF deployment will fail. As soon as we have separate build/deploy steps and use that on a CI/CD system, a missing check could lead to deployment errors.
My proposal is, that the validation should check that the same log stream subscription is only defined once - throughout all event definitions.

horike37 commented:
@buggy
Nice catch! I will support what you are writing!

@pmuens
Thanks for the review! fixed all what you pointed out!

@andymac4182 
Good idea! 

>We are doing this currently to push our lambda logs to our logging server.

How did you implement it? Cloud you give me more details?

@HyperBrain 
Thank you for the advice! That's make sense! I didn't understand this.
I will check this in the core code.

andymac4182 commented:
@horike37 

Inside of our handlers we are using https://github.com/structured-log/structured-log to console which outputs JSON and a human readable version. From there we have a single lambda function that receives all the logs and combines them into the right format for our logging system. It also pulls the lambda logs apart and turns them into logs we can use. https://gist.github.com/andymac4182/33ccf700fc25a29bce876a2943700a5d this is all sent via HTTP to Seq http://getseq.net which is our central logging system.

horike37 commented:
Hey @pmuens 
I have updated the following
- Support a short form configuration
```yaml
    events:
        - cloudwatchLog: '/aws/lambda/hello'
```
- Add duplication check for loggroup name as commented by @HyperBrain 

But I'm facing the following problems. I need your help :disappointed_relieved:

1. `AWS::Logs::SubscriptionFilter` resource doesn't support for `arn` to specify logGroup. Do you have any ideas for supported `arn` in this event?

2. There is a case which CF update fails. 

Steps to Reproduce 
- Write serverless.yml as follow, and run `sls deploy` at first
```yaml
functions:
  hello:
    handler: handler.hello
    events:
      - cloudwatchLog: '/aws/lambda/hello1'
  hello2:
    handler: handler.hello
    events:
      - cloudwatchLog: '/aws/lambda/hello2'
```
- Next, edit serverless.yml as follow, and run `sls deploy` again, so would be failed CF..
```yaml
functions:
  hello:
    handler: handler.hello
    events:
      - cloudwatchLog: '/aws/lambda/hello2'
  hello2:
    handler: handler.hello
    events:
      - cloudwatchLog: '/aws/lambda/hello1'
```

It seems to be due to conflict logGroup duplication problem during CF updates temporarily.
I'm considering to solve it that delete subscription filters from all Lambda in serverless.yml before CF updates. What do you think? Do you have a simpler idea?

HyperBrain commented:
For issue 2: Would it work if you set some dependencies on the subscriptionfilter resources? In CF you can add a DependsOn as array of logical resource ids to any resource (as sibling to the Properties property).
Maybe there is a combination that would let CF remove and create the filters in a defined order.
Can you try with making the filter dependent on the function or maybe in combination with the permission?

I do not have enough experience with the loggroup definition in CF to make any valid statement here. The only things I used up to now were the loggroup resources, which can be referenced by their logical id in the subscription. Do we really need a direct arn specification here? If the loggroup is maintained in a different CF stack you might be able to use Fn::ImportValue to fetch the reference cross stack - but that's only an assumption.

horike37 commented:
@HyperBrain 
Thank you for your help!
I tried some patterns with a DependsOn Today, but couldn't resolve it...
Since Lambdas are deployed in parallel through CF, its dependency might become very complicated if  subscriptionfilters exists across Lambdas.

pmuens commented:
Thanks for the updates @horike37 🥇 and thanks for chiming in @HyperBrain 👍 

> AWS::Logs::SubscriptionFilter resource doesn't support for arn to specify logGroup. Do you have any ideas for supported arn in this event?

Maybe we can leave that feature out for the first iteration or implement what @HyperBrain mentioned above. Also don't have that much experience on that end.

> I tried some patterns with a DependsOn Today, but couldn't resolve it...

🤔 could you please share some code snippets how the `DependsOn` looked like? Usually that should fix the problem.

horike37 commented:
Thanks for the reply @pmuens !

>Maybe we can leave that feature out for the first iteration or implement what @HyperBrain >mentioned above. Also don't have that much experience on that end.

Got it.

>could you please share some code snippets how the DependsOn looked like? Usually that should fix the problem.

I have tried out the two pattern. 

## Add ref to lambdaLogicalId

Added  "DependsOn" the reference  to Lambda
```
"DependsOn": ["${lambdaPermissionLogicalId}","${lambdaLogicalId}"],
```
Changed from the following line to the above.  
https://github.com/horike37/serverless/blob/add-cloudwatch-logs-event/lib/plugins/aws/deploy/compile/events/cloudWatchLog/index.js#L84

## Add ref to lambdaLogicalId and CloudwathcLog Resource name became randam

In addition to the above example, CloudwathcLog Resource name became randam.

Added `${(new Date()).getTime().toString()}` to the following line.
https://github.com/horike37/serverless/blob/add-cloudwatch-logs-event/lib/plugins/aws/lib/naming.js#L226

Please check it.

pmuens commented:
Hey @horike37 thanks for your response!

I reproduced the bug on my machine. The problem AFAIK is that you can only attach one subscription filter to one log group at any given time (see the last entry in this table: http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html).

In our case both log groups are assigned to listen to their respective Lambda functions on the first deployment. Once we change the order, CloudFormation tries to attach another SubscriptionFilter instead of removing the old one first and then attaching the new one. (I hope that this is understandable 😄 ).

So adding `DependsOn` might be hard here since we have to make sure that CloudFormation will issue a detachment before reattaching the subscription filters.

I'm not sure how we should proceed here. Maybe this is just an edge case and re-arranging subscription filters is not a common use case. In this case we could just move on and document this behavior...

I'll wrap my head around this the upcoming days and will get back here... 💭 🤔 

**Aside:** We had a similar problem with the `name` property for the `cloudwatchEvent` event source (https://github.com/serverless/serverless/pull/3102#issuecomment-273650342)

horike37 commented:
@pmuens 
>In our case both log groups are assigned to listen to their respective Lambda functions on the first >deployment. Once we change the order, CloudFormation tries to attach another SubscriptionFilter >instead of removing the old one first and then attaching the new one. (I hope that this is >understandable 😄 ).
Yes I see. 😄 CF temporarily try to attach the second loggroup, so it seems difficult to solve using `DependsOn`. 

It might be better that ask an AWS support way of solving once.
Or once removed all SubscriptionFilter just like the following comment
https://github.com/serverless/serverless/pull/3102#issuecomment-274022024

pmuens commented:
Ok. I agree. We should ping AWS about this problem since I don't see a way to resolve it via CloudFormation.

We should just document this behavior and then get this PR ready to be merged. We can still fix it in an upcoming release since this scenario might be an edge case anyway.

What do you think about that @horike37 ?

horike37 commented:
@pmuens 
I agree. It is OK to release the future before solving the problem on condition that would document this behavior.

pmuens commented:
@horike37 nice! Could you document this behavior in the `cloudwatchLog` docs? Otherwise if you don't have time I can do it.

horike37 commented:
@pmuens 
I will do it!
Where should I add it to the place in doc?
I think it would be better to add head or last of the sentence. 🤔
https://github.com/serverless/serverless/pull/3407/files#diff-b16d2768903aa0ebbfdcb3d7175d4ab9

pmuens commented:
@horike37 awesome. Thank you!

Yes, I think maybe a separate headline named "Current gotchas" or something like that will be sufficient.

```
## Current gotachs

Lorem ipsum...
```

Thanks for updating the docs! I'll do a deep dive into this PR tomorrow so that we can get it merged ASAP and have it ready for 1.11 💯

horike37 commented:
@pmuens 
Just added the attention message to doc. Please check it :bowtie:

pmuens commented:
Thanks @horike37 👍 Will do!

horike37 commented:
🎉🎉🎉🎉🎉:star::star:
